### PR TITLE
(#105) Fix Router.match pathAndMethod to only include layers that hav…

### PR DESCRIPTION
**Pull request by Gary-Osteen-Q2**
…e methods

### CODE REVIEW

**Review by dominicegginton — APPROVED** [review]

**Comment by dominicegginton** [comment]
@Gary-Osteen-Q2 Thanks for submitting a PR to fix #105. I have reviewed the changes. Just need a collaborator to review now. @niftylettuce do you have time to have a look at this? Let me know if there is anything I missed in my review, I am still new to this 😄

**Comment by Gary-Osteen-Q2** [comment]
Can anyone take some time to look this over?

**Review by jmealo — APPROVED** [review]
@Gary-Osteen-Q2: Congratulations on your first open source contribution. You included tests and these changes look good to me. I will be building against your branch while I work on some Open Telemetry improvements.

**Comment by jmealo** [comment]
@niftylettuce: I believe this PR may be required to address this upstream issue:
https://github.com/open-telemetry/opentelemetry-js-contrib/issues/222

**Comment by Gary-Osteen-Q2** [comment]
@jmealo Thanks! I've always looked forward to giving back to open source eventually. I am elated to think that it might also fix some upstream issues!

**Comment by niftylettuce** [comment]
Sorry I dropped the ball on this, is this ready to go? Can a few people look over this and ping me back?

**Comment by Gary-Osteen-Q2** [comment]
@niftylettuce it is ready to go from code side. I don't know what else would be needed.

**Review by 3imed-jaberi — COMMENTED** [review]
@Gary-Osteen-Q2, thank you for your efforts.
I think it's better to keep "path" over "pathOnly" as var/prop name. just for make the update smooth.

**Comment by Gary-Osteen-Q2** [comment]
@3imed-jaberi I'm not sure I follow how it would make the update smooth. Logically the layer that gets added to pathOnly have no methods to match in the first place. Whereas pathNotMethod the layer did have a method and it didn't match. Path would then be too generic to me as I would expect others to come in and think that the path matched everything.

**Comment by jmealo** [comment]
We should aim not to break this, which checks the `path`:
https://github.com/open-telemetry/opentelemetry-js-contrib/tree/master/plugins/node/opentelemetry-koa-instrumentation/src

**Comment by Gary-Osteen-Q2** [comment]
@jmealo pathOnly is used internally by Router.prototype.match which is a @private method. No outside project should be looking at that variable directly. The ctx.matched gets set at the same place as before in Router.prototype.routes. I don't see how this should affect any other dependent project. Inside Router.prototype.routes the path and ctx.path are unchanged.

**Comment by jmealo** [comment]
@Gary-Osteen-Q2: Sorry for my delayed reply. Your reasoning on a [`private` method](https://tc39.es/proposal-private-methods/) is correct in theory, however, the TC39 private methods and getter/setters proposal for JavaScript classes proposal is still a stage 3 draft. Typescript and JSDOC annotations can mark a method as private but JavaScript does not enforce it (yet).

As @3imed-jaberi: pointed out that this PR as implemented would need a `MAJOR` semver bump because it's an incompatible/breaking change.

I saw an actual monkey patch in the wild that wrapped internal methods of koa-router to set the operation name to the route on outgoing telemetry because there was no way to do it with the public API. npm is full of instrumentation libraries that monkey patch "private" methods.

Instrumentation developers can't really wait/lobby for each upstream projects to change the public API to expose its internals for what they'll consider a niche use case and their users needed better instrumentation yesterday. Since JavaScript doesn't have `private` methods, we have to tread lightly until there's a TypeScript/ES-Next native ecosystem. Ideally library authors will provide their own instrumentation with time. If that's what you're into you Deno is trying to be that thing.

**TLDR: @niftylettuce If we change the `path` back to `pathOnly` this is good to go with a `PATCH` release. If not, this needs to be a `MAJOR`.**

**Comment by niftylettuce** [comment]
@JacobMGEvans @Gary-Osteen-Q2 I think it is inferred that `path` is the same as `pathOnly` (from naming point of view) - is there any reason other than naming to have it renamed?  Or perhaps have both and add the new one and use that internally?  I would keep it backwards compatible despite the major semver change which will be shown in the changelog in releases tab of GH here.

**Review by 3imed-jaberi — APPROVED** [review]
LGTM in next bump release

**Comment by 3imed-jaberi** [comment]
@Gary-Osteen-Q2 can you please rebase with the master branch to resolve the conflict and go on here

**Comment by 3imed-jaberi** [comment]
@koa/router@15.0.0 has been published 🎉 and the issues this PR aims to fix are already resolved in v15. Thanks for your contribution!
